### PR TITLE
fix diagnostics showing up on file/dir with same prefix

### PR DESCRIPTION
**Pull request by chomosuke**
If A's path is B's prefix, then diagnostics in B will show up in both of them. For example, if there's diagnostics for `lua/nvim-tree.lua`, `lua/nvim-tree/` will also show diagnostics even though non of it's children have any.

This PR fixes that.

### CODE REVIEW

File: lua/nvim-tree/diagnostics.lua
Comment (alex-courtis) [inline lua/nvim-tree/diagnostics.lua:119]: Good catch!

Unfortunately we need to deal with windows powershell paths `C:\User\me`

Rather than messing around with path separators could we use something like `vim.fn.fnamemodify` `:h` and match on that?

string.find plain is also useful so that the separators aren't treated as magic characters.

**Review by alex-courtis — CHANGES_REQUESTED** [review]

File: lua/nvim-tree/diagnostics.lua
Comment (chomosuke) [reply lua/nvim-tree/diagnostics.lua:119]: The problem with `:h` is that `vim.fn.fnamemodify('lua/nvim-tree/diagnostics.lua', ':h')` will match `'lua/nvim'`. Though I do agree that `fnamemodify` is better and cleaner. I would be using `:p` instead which will add `/` if path is a directory.

**Comment by chomosuke** [comment]
Hmmmmm now that I look at it `vim.fn.fnamemodify` isn't really cleaner? I don't have a windows machine though so I don't know if all the `vim.fn.fnamemodify` are necessary and if `utils.canonical_path` are still necessary. I'd appreciate if someone with a windows machine can test that. I don't really have a strong preference between `fnamemodify` and `gsub` though so I'd leave it up to @alex-courtis to decide. Pls revert the last commit before you merge if you prefer `gsub` :).

**Comment by alex-courtis** [comment]
Unfortunately I am out of time; I'll get to this one on the weekend.

**Comment by alex-courtis** [comment]
> Hmmmmm now that I look at it `vim.fn.fnamemodify` isn't really cleaner? I don't have a windows machine though so I don't know if all the `vim.fn.fnamemodify` are necessary and if `utils.canonical_path` are still necessary. I'd appreciate if someone with a windows machine can test that. I don't really have a strong preference between `fnamemodify` and `gsub` though so I'd leave it up to @alex-courtis to decide. Pls revert the last commit before you merge if you prefer `gsub` :).

startswith is very useful. I should use that more.
Unfortunately it is interpreting a pattern:

```lua
print(vim.startswith("a/oeu", "a/o"))
-- true
print(vim.startswith("a\oeu", "a\o"))
-- invalid escape sequence near '"a'
```

**Comment by alex-courtis** [comment]
Looks like startswith is a winner: it doesn't match any vim or lua patterns. Much simpler than string.find plain.

```lua
  local needle = "/cc"
  local haystack = "/cc/dd/ee"

  log.line("dev", "needle '%s' haystack '%s' startswith   %s", needle, haystack, vim.startswith(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' vim.fn.match %s", needle, haystack, vim.fn.match(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' string.match %s", needle, haystack, string.match(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' string.find  %s", needle, haystack, string.find(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' string.find  %s", needle, haystack, string.find(haystack, needle, 1, true))

  log.line("dev", "")

  needle = "c:\\cc"
  haystack = "c:\\cc\\dd\\ee"

  log.line("dev", "needle '%s' haystack '%s' startswith   %s", needle, haystack, vim.startswith(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' vim.fn.match %s", needle, haystack, vim.fn.match(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' string.match %s", needle, haystack, string.match(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' string.find  %s", needle, haystack, string.find(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' string.find  %s", needle, haystack, string.find(haystack, needle, 1, true))

  log.line("dev", "")

  needle = "c:\\.c"
  haystack = "c:\\cc\\dd\\ee"

  log.line("dev", "needle '%s' haystack '%s' startswith   %s", needle, haystack, vim.startswith(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' vim.fn.match %s", needle, haystack, vim.fn.match(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' string.match %s", needle, haystack, string.match(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' string.find  %s", needle, haystack, string.find(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' string.find  %s", needle, haystack, string.find(haystack, needle, 1, true))

  log.line("dev", "")

  needle = "c:\\%ac"
  haystack = "c:\\cc\\dd\\ee"

  log.line("dev", "needle '%s' haystack '%s' startswith   %s", needle, haystack, vim.startswith(haystack, needle))
  -- log.line("dev", "needle '%s' haystack '%s' vim.fn.match %s", needle, haystack, vim.fn.match(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' string.match %s", needle, haystack, string.match(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' string.find  %s", needle, haystack, string.find(haystack, needle))
  log.line("dev", "needle '%s' haystack '%s' string.find  %s", needle, haystack, string.find(haystack, needle, 1, true))
```

```
[2022-12-23 11:56:13] [dev] needle '/cc' haystack '/cc/dd/ee' startswith   true
[2022-12-23 11:56:13] [dev] needle '/cc' haystack '/cc/dd/ee' vim.fn.match 0
[2022-12-23 11:56:13] [dev] needle '/cc' haystack '/cc/dd/ee' string.match /cc
[2022-12-23 11:56:13] [dev] needle '/cc' haystack '/cc/dd/ee' string.find  1
[2022-12-23 11:56:13] [dev] needle '/cc' haystack '/cc/dd/ee' string.find  1
[2022-12-23 11:56:13] [dev]
[2022-12-23 11:56:13] [dev] needle 'c:\cc' haystack 'c:\cc\dd\ee' startswith   true
[2022-12-23 11:56:13] [dev] needle 'c:\cc' haystack 'c:\cc\dd\ee' vim.fn.match -1
[2022-12-23 11:56:13] [dev] needle 'c:\cc' haystack 'c:\cc\dd\ee' string.match c:\cc
[2022-12-23 11:56:13] [dev] needle 'c:\cc' haystack 'c:\cc\dd\ee' string.find  1
[2022-12-23 11:56:13] [dev] needle 'c:\cc' haystack 'c:\cc\dd\ee' string.find  1
[2022-12-23 11:56:13] [dev]
[2022-12-23 11:56:13] [dev] needle 'c:\.c' haystack 'c:\cc\dd\ee' startswith   false
[2022-12-23 11:56:13] [dev] needle 'c:\.c' haystack 'c:\cc\dd\ee' vim.fn.match -1
[2022-12-23 11:56:13] [dev] needle 'c:\.c' haystack 'c:\cc\dd\ee' string.match c:\cc
[2022-12-23 11:56:13] [dev] needle 'c:\.c' haystack 'c:\cc\dd\ee' string.find  1
[2022-12-23 11:56:13] [dev] needle 'c:\.c' haystack 'c:\cc\dd\ee' string.find  nil
[2022-12-23 11:56:13] [dev]
[2022-12-23 11:56:13] [dev] needle 'c:\%ac' haystack 'c:\cc\dd\ee' startswith   false
[2022-12-23 11:56:13] [dev] needle 'c:\%ac' haystack 'c:\cc\dd\ee' string.match c:\cc
[2022-12-23 11:56:13] [dev] needle 'c:\%ac' haystack 'c:\cc\dd\ee' string.find  1
[2022-12-23 11:56:13] [dev] needle 'c:\%ac' haystack 'c:\cc\dd\ee' string.find  nil
```

**Comment by alex-courtis** [comment]
> Hmmmmm now that I look at it `vim.fn.fnamemodify` isn't really cleaner?

The key part is the `node.nodes` test.

gsub works even for crazy paths like `/tmp/bleh\bleh`

**Review by alex-courtis — APPROVED** [review]
Tested OK:
`nvim-tree.lua/lua` show_on_dirs show_on_open_dirs
* false false
* true false
* false true
* true true

`nvim-tree/lua\lua`
as above

**Pull request merged**
